Add BigQueryRpc default implementation

mziccard opened the pull request:
This PR adds the `DefaultBigQueryRpc` class to the BigQuery module.
Base functionalities are implemented. Possibly more will come later.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,421 @@
+
+  public static final String DEFAULT_PROJECTION = "full";
+  // see: https://cloud.google.com/bigquery/troubleshooting-errors
+  private static final Set<Integer> RETRYABLE_CODES = ImmutableSet.of(503);
```
Should we add [these](https://cloud.google.com/bigquery/loading-data-post-request#best-practices) other 500-level error codes?

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,421 @@
+
+  private static BigQueryException translate(GoogleJsonError exception) {
+    boolean retryable = RETRYABLE_CODES.contains(exception.getCode())
+        || "InternalError".equals(exception.getMessage());
```
I think this line should be `|| "internalError".equals(exception.getReason());` because [this link](https://cloud.google.com/bigquery/troubleshooting-errors) (second line under "Error table") says the "error code" (which looks like it's camel case) maps to the error's reason.

mziccard replied:
Good catch thanks!

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,421 @@
+  }
+
+  @Override
+  public Dataset patch(Dataset dataset, Map<Option, ?> options) throws BigQueryException {
```
Should `patch` be renamed `update` as per #321?

mziccard replied:
I thought that issue only applied to `XXXService` classes and that in `XXXRpc` we should be keeping the name of the rpc. I might have got it wrong though.

ajkannan replied:
Ah, you're right, ignore that comment and the one below.

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,421 @@
+  }
+
+  @Override
+  public Table patch(Table table, Map<Option, ?> options) throws BigQueryException {
```
Like above, should patch --> update?

ajkannan commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,421 @@
+      Iterable<DatasetList.Datasets> datasets = datasetsList.getDatasets();
+      return Tuple.of(datasetsList.getNextPageToken(),
+          Iterables.transform(datasets != null ? datasets : ImmutableList.<DatasetList.Datasets>of(),
+              new Function<DatasetList.Datasets, Dataset>() {
```
It's interesting that DatasetList.Datasets returns only partial information about the dataset.  We should probably document that if the user wants other information about the dataset not included in the dataset reference, that they should call "get."

mziccard commented:
Thanks @ajkannan for the pass. I fixed your comment and fixed a bug when `null` jobs/tables lists are returned.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,448 @@
+
+  private static BigQueryException translate(IOException exception) {
+    BigQueryException translated;
+    if (exception instanceof GoogleJsonResponseException
```
This logic is a common pattern. Would be nice to put it in `BaseServiceException` though this could be done as a separate PR

mziccard replied:
Right, opened #413 for this

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,448 @@
+      return bigquery.datasets()
+          .get(this.options.projectId(), datasetId)
+          .setFields(FIELDS.getString(options))
+          .setQuotaUser(QUOTA_USER.getString(options))
```
QUOTA_USER and USER_IP are not BigQuery specific but rather part of the [api client](https://developers.google.com/console/help/new/#cappingusage).

Therefore it is also available to other services such as GCS. So far, we ignored it.
Do you think it is necessary/usefull?

mziccard replied:
I added them as I was trying to have a comprehensive understanding of all possible options for each method. Since we ignored them in other services we can probably remove support for them (unless you want to add them to other modules - I don't mind doing it).

aozarov replied:
Let's remove them for now.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,448 @@
+      bigquery.datasets().delete(this.options.projectId(), datasetId)
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .setDeleteContents(DELETE_CONTENTS.getBoolean(options))
```
is there a specific error code for the case of `false` and there is data?

mziccard replied:
400 is returned in that case, with message:

> Dataset \<dataset\> is still in use

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,448 @@
+  @Override
+  public Dataset patch(Dataset dataset, Map<Option, ?> options) throws BigQueryException {
+    try {
+      DatasetReference reference = dataset.getDatasetReference();
```
So, one can patch a `Dataset` for a different project?

mziccard replied:
Yes, same holds for Table.
I realized that in create methods I take the project's id from `BigQueryOption` while in patch I take it from `DatasetReference`, this might be quite an inconsistency so we should decide the way to go.

1. If we take the project id from `BigQueryOption` the request will fail if `dataset.getDatasetReference().getProjectId()` does not match
2. If we take the project id from `DatasetReference` we allow patching/creating for different projects. However, we do not support get for different projects as `getDataset(String datasetId)` takes the project id from `BigQueryOption`

I prefer option 1), do you have any thoughts?

aozarov replied:
I also prefer option 1.

aozarov commented on gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java:
```diff
@@ -0,0 +1,448 @@
+          .setPageToken(PAGE_TOKEN.getString(options))
+          .setStartIndex(START_INDEX.getLong(options) != null ?
+              BigInteger.valueOf(START_INDEX.getLong(options)) : null)
+          .setTimeoutMs(TIMEOUT.getLong(options))
```
In the API we will need to find a better name for that.

aozarov commented:
Looks good. sorry that review took so long.
Few questions and one item to think about (QUOTA_USER and USER_IP).

mziccard commented:
I fixed the comments:
- Removed quota option
- Always use `options.projectId` to set project's id
- Removed a 404 I missed
- Fixed a bug with job's id
